### PR TITLE
8289644: Add a throwIf method to RuntimeException

### PR DESCRIPTION
**Pull request by davidalayachew**
Adding a new method to RuntimeException that will allow it to concisely validate preconditions. This is meant to be similar functionality to Objects.requireNonNull().

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Error
&nbsp;⚠️ OCA signatory status must be verified

### Issue
 * [JDK-8289644](https://bugs.openjdk.org/browse/JDK-8289644): Add a throwIf method to RuntimeException

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9355/head:pull/9355` \
`$ git checkout pull/9355`

Update a local copy of the PR: \
`$ git checkout pull/9355` \
`$ git pull https://git.openjdk.org/jdk pull/9355/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9355`

View PR using the GUI difftool: \
`$ git pr show -t 9355`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9355.diff">https://git.openjdk.org/jdk/pull/9355.diff</a>

</details>

### CODE REVIEW

**Comment by davidalayachew** [comment]
Ok, I have submitted a RFE through the Java Bug Database. I suppose I need to wait on that to be approved then? And then I change this pull request title to include the id number followed by a colon and my original title?

**Comment by davidalayachew** [comment]
The internal id for that RFE is 9073547. It will turn into a public RFE ID later, but in case that facilitates processing, let me put it here.

**Comment by davidalayachew** [comment]
Turned this into a draft until the public RFE ID is given to me, which allow my pull request to follow the rules.

**Review by kristylee88 — APPROVED** [review]

**Review by kristylee88 — CHANGES_REQUESTED** [review]
It failed

**Comment by davidalayachew** [comment]
Hello kristylee88, and thank you for approving my changes. The history says that you suggested changes, but I am unable to see them.

**Comment by davidalayachew** [comment]
Also, I see your comment about it failed. I believe I am waiting on my OCA to be approved. The Oracle person I spoke to over email told me to make my pull request, and then that way, they could finish my OCA approval process.

**Comment by davidalayachew** [comment]
Are there any immediate steps I should take so that I can pass jcheck? Or am I just waiting on my OCA then everything should be good?

**Review by kristylee88 — APPROVED** [review]

**Comment by davidalayachew** [comment]
I see that you approved the changes again @kristylee88. I have marked my pull request as ready for review. Please let me know if there is anything I need to do or undo.

This is my first time doing any sort of pull request, let alone contributing to OpenJDK, so I am a bit lost and confused on the process. Apologies if I have made any mistakes. Please let me know if I need to redo or I need to undo stuff.

**Comment by TheShermanTanker** [comment]
Hey @davidalayachew,

Thanks a lot for contributing to OpenJDK! It may take some time for the contributor agreement to be approved, so all you can do is wait until you get the go ahead unfortunately. While you're waiting though, I have some advice for you:

- The change you are implementing here is a non-trivial one that's also in a very widely used section of the Java Platform as a whole; The bar for a Pull Request being accepted into core-libs is significantly higher than something like, say, build-dev (Which deals with the internal build system). For things like these, it's better to communicate with the relevant development group first about your idea, and see if it can be improved or if it 's a good addition, rather than showing up with changes directly without any warning, as it can cause issues if they aren't as helpful to Java as you would've expected.
- You need an entry in the tracker for Skara (OpenJDK's own version control automation system) to notify the relevant maintainers about this PR, otherwise it'll never be seen as most of the Reviewers don't typically scan the Pull Request section. I could create one for you, but for the reasons I mentioned above, it'd be much better for you to communicate with the mailing lists about your new idea first. Trust me when I say that discussing a solution with the appropriate team first is better than having your PR publicly rejected on this repository (Even more so considering some groups - Such as core-libs - Have members that are known to be particularly harsh when commenting on why a certain PR shouldn't be accepted!)
- I'm not a core-libs developer, so I can't really comment on these changes, but they don't appear to be correct to me at first glance. Like I said though, I'm not involved in this area of Java, so don't take my word for it.

Since I'm only an Author in the project, I unfortunately can't approve your signed agreement. The only way I can help you, at least for now, is by contacting the mailing lists asking for a review of this PR and/or for approval of your agreement, so do let me know if you want me to do that.

As for jcheck, the reason it's failing is due to Skara not being able to resolve your contributor status, and unfortunately, as I said above, all you can do is really wait. Who was the Oracle person you spoke to? For all you know, they could be an active member on OpenJDK that can be easily reached on the mailing lists.

All the best in your contributions to OpenJDK!
~Julian

**Comment by davidalayachew** [comment]
Thank you for the guidance and advice @TheShermanTanker. From what you are describing, it sounds like my best bet would be to delete this pull request, reach out to the relevant parties about my change, discuss with them about what I want to do, and then once I get perspective/feedback/criticism/approval from them, then I make my pull request?

As for the Oracle person I was speaking to, their name was Oleksandra Pavlusieva.

Lastly, you mentioned that the change doesn't seem correct. Could you give me some feedback on what part does not seem right? I fully intend to communicate with the mailing list like you said, but it might be good for me if I can see my solutions obvious flaws, so if you see something wrong with my idea, I would appreciate the feedback.

Thank you for your help!

**Comment by TheShermanTanker** [comment]
Since you've already made the PR, there wouldn't be much of a point in deleting it. It would be a good idea however to send a mail to [core-libs-dev@openjdk.org](mailto:core-libs-dev@openjdk.org) to see where to go from here. Like I said, I can't really give much of a helpful review since this isn't an area of Java that I work on, but to my knowledge an entirely new method added to something in the core libraries is generally frowned upon, unless there isn't a better solution. I could be entirely wrong here, but I also do vaguely recall seeing throwIf somewhere else, but don't take my word for it, like I said I could be completely mistaken.

When you do get feedback and want to address it with new changes though, try not to force push any new commits; Use merges instead, it makes the actual reviewer's lives somewhat easier.

All the best!

**Comment by davidalayachew** [comment]
I am typing up my email to them now, thank you for very much for the guidance.

And thank you for the feedback. I will be more prepared to answer that question when the mailing list folks review my suggestion.

I am afraid I don't understand your last paragraph though. My plan was to get feedback, and then make a commit to the branch on my fork of the jdk, then it would show up here as a third commit. Is that a correct way of going about things, or are you suggesting I do it a different way? An example of me doing this is that, when I first made this change, I had extra whitespace, and jcheck failed me for that. I then committed a change to my fork, and then I passed that stage of jcheck (even though it still failed because of OCA). Should I do it like that again when I want to make changes to my code?

Thank you again for all of your help!

**Comment by davidalayachew** [comment]
Also, I pulled down the jdk and did a grep search for the phrase "hrowIf". Found close to 50 results, but they were all trying to make their own precondition methods for their own class (example - the StructuredTaskScope.java -- ShutdownOnFailure#throwIfFailed). Hopefully this means I'm good and that I am not recreating functionality that already exists for folks.

**Comment by TheShermanTanker** [comment]
> I am typing up my email to them now, thank you for very much for the guidance.
> 
> And thank you for the feedback. I will be more prepared to answer that question when the mailing list folks review my suggestion.
> 
> I am afraid I don't understand your last paragraph though. My plan was to get feedback, and then make a commit to the branch on my fork of the jdk, then it would show up here as a third commit. Is that a correct way of going about things, or are you suggesting I do it a different way? An example of me doing this is that, when I first made this change, I had extra whitespace, and jcheck failed me for that. I then committed a change to my fork, and then I passed that stage of jcheck (even though it still failed because of OCA). Should I do it like that again when I want to make changes to my code?
> 
> Thank you again for all of your help!

No no, what I meant is when you push further commits to your fork, you shouldn't `git push origin branch-name --force`, instead if the branch is behind the upstream, just `git merge` it instead

**Comment by davidalayachew** [comment]
Ok, thank you very much. I actually am attempting to improve my solution before reaching out to the team, so I will put that to use now.

Also, please ignore any excess notifications about branches being deleted or restored. I am not fluent in Git(Hub), so it's taking a few iterations to get things to go the way I want them to.

**Comment by davidalayachew** [comment]
@TheShermanTanker apologies for depending on you so much. Would this be a proper example of a merge in a way that doesn't disrupt reviewers?

https://github.com/openjdk/jdk/pull/9355/commits/c9ad46f267ba54ac2264a364b22b699ae6c36d3b

You'll have to forgive me for asking, I have not been using the command line for this thus far, so I am trying to see if this is an acceptable alternative before attempting to get my command line git working.

File: src/java.base/share/classes/java/lang/RuntimeException.java
Comment (liach) [inline src/java.base/share/classes/java/lang/RuntimeException.java:129]: `IllegalArgumentException.throwIf(3 > 2)`
Uh, this method you added is an instance method, and you can't call it like that.

In general, you should avoid creating exception instances if you don't need to throw them, since filling in stack trace is extremely costly.

Comment (davidalayachew) [reply src/java.base/share/classes/java/lang/RuntimeException.java:129]: Terribly sorry, I meant to do it like this.

    new IllegalArgumentException.throwIf(3 > 2);

Thank you for pointing that out.

But yes, your point still stands - it is expensive to create stack traces, so therefore, this is not an "equivalent, but more concise" solution, it is significantly more costly in exchange for some readability.

Perhaps it actually would work better as a static method? Then, the stack trace is only created upon breaking the precondition.

Thank you again for your feedback.

Comment (davidalayachew) [reply src/java.base/share/classes/java/lang/RuntimeException.java:129]: Something like this.

```
public static void throwIf(boolean throwGuard, String message) {
    if (throwGuard) {
        throw new RuntimeException(message);
    }
}
```

**Comment by TheShermanTanker** [comment]
> 

Yep, merge looks good.

File: src/java.base/share/classes/java/lang/RuntimeException.java
Comment (liach) [reply src/java.base/share/classes/java/lang/RuntimeException.java]: I recommend you learn the Java Programming Language, such as how static methods work, before making contributions to the OpenJDK. See section 5 of https://openjdk.org/contribute/

Comment (TheShermanTanker) [reply src/java.base/share/classes/java/lang/RuntimeException.java]: > I recommend you learn the Java Programming Language, such as how static methods work, before making contributions to the OpenJDK. See section 5 of https://openjdk.org/contribute/

I believe the PR author really means something like
```
IllegalArgumentException exception = new IllegalArgumentException();
exception.throwIf(...);
```
where throwIf is called on an instance of an exception, instead of a check in an if block as per usual. Whether this is a good solution compared to the usual approach I can't comment on, but I think there might be a bit of misunderstanding involved

Comment (davidalayachew) [reply src/java.base/share/classes/java/lang/RuntimeException.java]: @TheShermanTanker, I believe the reason why they made that comment was because my static function suggestion does not satisfy the goals I originally claimed this pull request would satisfy.

Looking at my static function suggestion, the method is no longer specialized for each Exception class. Yes, I can do `IllegalArgumentException.throwIf(true)`, but I will throw a RuntimeException, not an IllegalArgumentException.

I did not clarify that I was moving the goal posts, so I can understand if it came off like I thought a static method would somehow accomplish the original goals that I had set for myself. I think that that is why they suggested that I learn the language and how static methods work within it.

@liach, please feel free to correct me if I am misrepresenting you.

**Comment by davidalayachew** [comment]
@TheShermanTanker thank you for teaching me. I understand this process much better, and I feel well prepared for my next pull request. This is my first attempt at a pull request, so I appreciate your patience.

File: src/java.base/share/classes/java/lang/RuntimeException.java
Comment (liach) [reply src/java.base/share/classes/java/lang/RuntimeException.java]: Indeed, static methods won't work and you understand the reason.

I suggest you look at a Preconditions class in jdk.internal.util package for how jdk handles common exceptions. I suggest writing personal utilities or using library utilities if there is a particular type of IAE that happens often to you than adding this somewhat useless method, only good for it squashes code to one possibly too long line.

Comment (davidalayachew) [reply src/java.base/share/classes/java/lang/RuntimeException.java]: Thank you for responding and clarifying @liach. Your patience is appreciated.

I believe I have found the file you are suggesting. I attempted to use this class in my code, but it does not work. Likely because it is an internal jdk package. Regardless, this gives a template on how to create Precondition checks for myself. 

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/util/Preconditions.java

This definitely makes sense. By using a java.util.function.* class, I can defer object creation (or in this case, exception creation) until the moment that I use it, allowing me to bypass the concern you raised before. Alternatively, I could just create a custom method that does what I need it to, which segues into your point.

Based off what you are saying, my feature is something that doesn't provide enough value to warrant its inclusion. I understand what you are saying and it makes sense. Based off of that, it likely would be best for me to close this pull request and focus on making something else.

But to get a feel of the mindset of JDK Contributors, I'd like to ask a hypothetical question. If exception creation was not expensive, would this feature make sense to include in the JDK? I want to understand the level of quality I must aim for before making my next pull request to the JDK.

Comment (davidalayachew) [reply src/java.base/share/classes/java/lang/RuntimeException.java]: > _But to get a feel of the mindset of JDK Contributors, I'd like to ask a hypothetical question. If exception creation was not expensive, would this feature make sense to include in the JDK? I want to understand the level of quality I must aim for before making my next pull request to the JDK._

@TheShermanTanker and @kristylee88, I'd like to direct this question to you 2 too. I am new and learning, and I would like to come out of this with as much knowledge as I can about how to do better. If not that, then any other suggestions you all have to give would be greatly appreciated.

**Pull request closed without merging**
